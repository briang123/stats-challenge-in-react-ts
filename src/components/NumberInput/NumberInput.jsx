import React from 'react';
import { useNumberInput } from '../../hooks';
import styled from 'styled-components';

// interface IProps {
//   testid: string;
//   tabIndex: number;
//   placeholder: string;
//   value: number;
//   autoFocus: boolean;
//   setValue: (value: string) => void;
//   onSubmit: (e: React.FormEvent<HTMLButtonElement | HTMLInputElement>) => void;
// }

export const NumberInput = (
  { testid, tabIndex = -1, placeholder = '', value, autoFocus = false, setValue, onSubmit },
  ref
) => {
  const [onKeyUp, onKeyPress, onChange] = useNumberInput({
    setValue,
    onSubmit,
  });

  if (!setValue || !onSubmit) {
    console.error('setValue and onSubmit functions are required');
    return;
  }

  if (!ref) {
    console.error('This NumberInput component expects a ref, which is missing');
    return;
  }

  return (
    <StyledInput
      ref={ref}
      data-testid={testid}
      tabIndex={tabIndex}
      type="number"
      placeholder={placeholder}
      onKeyUp={onKeyUp}
      onKeyPress={onKeyPress}
      onChange={onChange}
      value={value}
      autoFocus={autoFocus}
    />
  );
};

// export default NumberInput;
export default React.forwardRef(NumberInput);
// export default React.forwardRef<HTMLInputElement, IProps>(NumberInput);

const StyledInput = styled.input`
  height: 75px;
  width: 70%;
  border-radius: 20px 0px 0px 20px;
  border: none;
  font-size: calc(20px + 2vmin);
  text-align: center;
  :focus {
    outline: none;
  }
  ::-webkit-input-placeholder {
    color: var(--disabledbtn);
  }
  ::-moz-placeholder {
    color: var(--disabledbtn);
  }
  :-ms-input-placeholder {
    color: var(--disabledbtn);
  }
  :-moz-placeholder {
    color: var(--disabledbtn);
  }
  ::-webkit-outer-spin-button,
  ::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  -moz-appearance: textfield;
`;
