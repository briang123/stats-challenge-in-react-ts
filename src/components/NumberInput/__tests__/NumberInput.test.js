import React from 'react';
import { cleanup, render, fireEvent } from '@testing-library/react';
import { create } from 'react-test-renderer';
import NumberInput from '../NumberInput';

afterEach(() => {
  cleanup();
  console.error.mockClear();
});

console.error = jest.fn();
const mockFn = jest.fn();

describe('<NumberInput />', () => {
  it('Input component without passing a function to the setValue or onSubmit props should log error and not render input', () => {
    const { queryByTitle } = render(<NumberInput />);
    expect(console.error).toHaveBeenCalledTimes(1);

    const input = queryByTitle('INPUT');
    expect(input).toBeNull();
  });

  it('Input component passing a function to the setValue and prop, but not passing a reference input should log error and not render input', () => {
    const { queryByTitle } = render(<NumberInput setValue={mockFn} onSubmit={mockFn} />);

    expect(console.error).toHaveBeenCalledTimes(1);

    const input = queryByTitle('INPUT');
    expect(input).toBeNull();
  });

  it('Input component passing an input ref and a function to both setValue and onSubmit props should not log error and render number input with tabindex = -1', () => {
    const ref = React.createRef();
    const { debug } = render(<NumberInput ref={ref} setValue={mockFn} onSubmit={mockFn} />);

    expect(console.error).not.toHaveBeenCalled();

    debug();

    expect(ref.current).not.toBeNull();
    expect(ref.current.getAttribute('type')).toEqual('number');
    expect(ref.current.getAttribute('tabindex')).toEqual('-1');
  });

  it('Input component passing ref and funcs should and placeholder of "foo" should render a placeholder of "foo" on the input', () => {
    const ref = React.createRef();
    render(<NumberInput ref={ref} setValue={mockFn} onSubmit={mockFn} placeholder="foo" />);

    expect(ref.current.getAttribute('placeholder')).toEqual('foo');
  });

  it('Input should remain empty if user enters non-numbers', () => {
    const ref = React.createRef();
    render(<NumberInput ref={ref} setValue={mockFn} onSubmit={mockFn} placeholder="foo" />);

    const input = ref.current;
    expect(input.value).toBe('');
    fireEvent.change(input, { target: { value: 'foo' } });
    expect(input.value).toBe('');
  });

  it('Input should accept only numbers', () => {
    const ref = React.createRef();
    render(<NumberInput ref={ref} setValue={mockFn} onSubmit={mockFn} placeholder="foo" />);

    const input = ref.current;
    expect(input.value).toBe('');
    fireEvent.change(input, { target: { value: '1' } });
    expect(input.value).toBe('1');
  });

  it('Should match snapshot with props', () => {
    const ref = React.createRef();
    const tree = create(
      <NumberInput ref={ref} setValue={mockFn} onSubmit={mockFn} placeholder="foo" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
