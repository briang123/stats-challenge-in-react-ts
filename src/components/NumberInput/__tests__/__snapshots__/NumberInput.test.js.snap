// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberInput /> Should match snapshot with props 1`] = `
.c0 {
  height: 75px;
  width: 70%;
  border-radius: 20px 0px 0px 20px;
  border: none;
  font-size: calc(20px + 2vmin);
  text-align: center;
  -moz-appearance: textfield;
}

.c0:focus {
  outline: none;
}

.c0::-webkit-input-placeholder {
  color: var(--disabledbtn);
}

.c0::-moz-placeholder {
  color: var(--disabledbtn);
}

.c0:-ms-input-placeholder {
  color: var(--disabledbtn);
}

.c0:-moz-placeholder {
  color: var(--disabledbtn);
}

.c0::-webkit-outer-spin-button,
.c0::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

<input
  autoFocus={false}
  className="c0"
  onChange={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  placeholder="foo"
  tabIndex={-1}
  type="number"
/>
`;
